packages/financial-templates-lib/src/price-feed/DefiPulseTotalPriceFeed.js

```diff
@@ -0,0 +1,144 @@
+    };
+  }
+
+  getCurrentPrice() {
+    return this.currentPrice;
+  }
+
+  getHistoricalPrice(time) {
+    if (this.lastUpdateTime === undefined) {
+      return undefined;
+    }
+
+    let closestTime = { timestamp: 0, tvlUSD: 0 };
+
+    // go through all values and find time that that is the largest and still less than 'time'
+    for (let i = 0; i < this.historicalPrices.length; i++) {
+      let past = this.historicalPrices[i].timestamp;
+      let val = this.historicalPrices[i].tvlUSD;
+
+      if (past > closestTime.timestamp && past < time) {
+        closestTime.timestamp = past;
+        closestTime.tvlUSD = val;
+      }
+    }
+
+    const historicalPrice = this.web3.utils.toWei((closestTime.tvlUSD / 1000000000).toFixed(6), "ether");
+
+    if (closestTime.timestamp === 0) {
+      return undefined;
+    } else {
+      return this.toBN(historicalPrice);
+    }
+  }
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  async update() {
+    const currentTime = this.getTime();
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== undefined && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
+      this.logger.debug({
+        at: "DEFI_PULSE_TOTAL_TVL_PriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTimes + this.minTimeBetweenUpdates - currentTime
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "DEFI_PULSE_TOTAL_TVL_PriceFeed",
+      message: "Updating",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime
+    });
+
+    // 1. Construct URLs.
+    const priceUrl = `https://data-api.defipulse.com/api/v1/defipulse/api/GetHistory?period=1w&api-key=${this.apiKey}`;
+
+    // 2. Send requests.
+    const response = await this.networker.getJson(priceUrl);
+
+    // 3. Check responses.
+    if (!response) {
+      throw new Error(`🚨Could not parse price result from url ${priceUrl}: ${JSON.stringify(response)}`);
+    }
+
+    // 4. Parse results.
+    // Return data structure:
+    //        [{
+    //            "timestamp":"1611511200"
+    //            "tvlUSD":25583565042,
+    //             ...
+    //          },
+    //           {
+    //            "timestamp":"1611507600",
+    //            "tvlUSD":25177860561
+    //             ...
+    //          },
+    //        ]
+
+    // Get tvlUSD for most most recent timestamp
+    let mostRecent = { timestamp: 0, tvlUSD: 0 };
+    for (let i = 0; i < response.length; i++) {
+      if (Number(response[i].timestamp) > mostRecent.timestamp) {
+        mostRecent.timestamp = Number(response[i].timestamp);
+        mostRecent.tvlUSD = Number(response[i].tvlUSD);
+      }
+    }
+
+    const newPrice = this.web3.utils.toWei((mostRecent.tvlUSD / 1000000000).toFixed(6), "ether");
```

Comment:
Can you provide a comment on what's happening here?

Also, do you think this should be a class function since you do the same operation in the getHistoricalPrice method?